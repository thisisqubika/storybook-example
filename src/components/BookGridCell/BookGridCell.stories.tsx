import type { Meta, StoryObj } from '@storybook/react';
import { ThemeWrapper } from '../../common/themes';
import { BookGridCell, type BookGridCellProps } from '.';
import { searchBooks_ReactWeb } from '../../../test/fixtures/openLibrary';

export default {
  component: BookGridCell,
} satisfies Meta<BookGridCellProps>; // meta

const storyData = searchBooks_ReactWeb.docs.find(({
  author_name, cover_i, language, publish_date, publisher, title,
}) => (
  !!author_name && !!cover_i && !!language && !!publish_date && !!publisher && !!title
)) ?? searchBooks_ReactWeb.docs[1];

export const Base: StoryObj<BookGridCellProps> = {
  args: {
    book: storyData,
  } satisfies BookGridCellProps,

  render(props: BookGridCellProps) {
    return (
      <ThemeWrapper>
        <BookGridCell {...props} />
      </ThemeWrapper>
    );
  },
}; // const Base
