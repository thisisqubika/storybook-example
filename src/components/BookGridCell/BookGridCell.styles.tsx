import { SxProps, Theme } from "@mui/material";

export function sxStyles(): Record<string, SxProps<Theme>> {
  return {
    card: {
      display: 'grid',
      gridTemplateColumns: '2fr 1fr',
      padding: '1rem',
    }, // card

    column1: {
      display: 'flex',
      flexDirection: 'column',
      gap: '2px',
    }, // column1

    column2: {
      alignItems: 'center',
      display: 'flex',
      justifyContent: 'center',
    } // column2
  };
}; // sxStyles
