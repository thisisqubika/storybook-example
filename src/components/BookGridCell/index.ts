export {
  default as BookGridCell,
  type BookGridCellProps,
} from './BookGridCell';
