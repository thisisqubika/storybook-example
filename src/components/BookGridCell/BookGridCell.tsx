import { Box, Card, Typography } from '@mui/material';
import { sxStyles } from './BookGridCell.styles';
import { OpenLibraryBook } from '../../services/openLibrary.types';
import openLibraryApi from '../../services/openLibrary';
  
export interface BookGridCellProps {
  book: OpenLibraryBook;
} // interface BookGridCellProps

function splitTitle(title: string) {
  const parts = title.split(/\s*[:-]\s+/);
  return parts[0] ? parts : [title];
}

function abbreviate(
  list: string[], etc?: string, maxLength = 3,
) {
  if (list.length > maxLength) {
    return list.slice(0, maxLength)
      .concat([etc ?? ` and ${list.length - maxLength} more`]);
  }
  return list;
} // abbreviate

export default function BookGridCell({
  book,
}: BookGridCellProps) {
  const {
    author_name: authorNames,
    cover_i: coverImageId,
    language: languages,
    publish_date: publicationDates,
    publisher: publishers,
    title,
  } = book;
  const styles = sxStyles();

  const [realTitle, subtitle] = splitTitle(title);

  const cardTitle = (
    <Typography data-testid='BookGridCell.title' variant='h4'>
      {realTitle}
    </Typography>
  );
  const cardSubtitle = subtitle ? (
    <Typography data-testid='BookGridCell.subtitle' variant='subtitle1'>
      {subtitle}
    </Typography>
  ) : null;

  const cardAuthors = (
    <Typography variant='body1'>
      {!authorNames ? 'Authors not available'
        : `By: ${abbreviate(authorNames, 'et al.').join(', ')}`}.
    </Typography>
  );
  const cardLanguages = (
    <Typography variant='body1'>
      {!languages ? 'Languages not available'
        : `Languages: ${abbreviate(languages).join(', ')}`}.
    </Typography>
  );
  const cardPublications = (
    <Typography variant='body1'>
      {!publishers ? `Publications data not available`
        : `Published by ${abbreviate(publishers).join('; ')} at ${
        !publicationDates ? 'dates not available' 
          : abbreviate(publicationDates).join(', ')}`
      }.
    </Typography>
  );

  const cardImage = coverImageId ? (
    <img
      alt='Books cover'
      data-testid='BookGridCell.img'
      src={`${openLibraryApi.bookCoverImage(coverImageId)}`}
    />
  ) : (
    <Typography variant='caption'>Image not avaiable</Typography>
  );

  return (
    <Card
      data-testid='BookGridCell'
      sx={styles.card}
    >
      <Box sx={styles.column1}>
        {cardTitle}
        {cardSubtitle}
        {cardAuthors}
        {cardLanguages}
        {cardPublications}
      </Box>
      <Box sx={styles.column2}>
        {cardImage}
      </Box>
    </Card>
  );
} // function BookGridCell
