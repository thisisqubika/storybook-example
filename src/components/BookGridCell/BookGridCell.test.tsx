import { describe, expect, test } from 'vitest';
import { composeStories } from "@storybook/react";
import { render } from '@testing-library/react';
import * as stories from './BookGridCell.stories';
import { type BookGridCellProps } from './BookGridCell';

const {
  Base: BookGridCellBaseStory,
} = composeStories(stories);

describe('BookGridCell', () => {
  test('stories match snapshot', () => {
    for (const Story of [
      BookGridCellBaseStory,
    ]) {
      const props = Story.args as BookGridCellProps;
      const { container } = render(
        <Story {...props} />
      );
      expect(container).toMatchSnapshot();
    }
  }); // test

}); // describe 'BookGridCell'
