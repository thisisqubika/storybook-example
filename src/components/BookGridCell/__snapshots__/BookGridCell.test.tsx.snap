// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BookGridCell > stories match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-dp0tfs-MuiPaper-root-MuiCard-root"
    data-testid="BookGridCell"
  >
    <div
      class="MuiBox-root css-3a3hw0"
    >
      <h4
        class="MuiTypography-root MuiTypography-h4 css-j95l75-MuiTypography-root"
        data-testid="BookGridCell.title"
      >
        React Application Architecture for Production
      </h4>
      <p
        class="MuiTypography-root MuiTypography-body1 css-e773oc-MuiTypography-root"
      >
        By: Alan Alickovic
        .
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-e773oc-MuiTypography-root"
      >
        Languages: eng
        .
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-e773oc-MuiTypography-root"
      >
        Published by Packt Publishing, Limited; Packt Publishing at 2023
        .
      </p>
    </div>
    <div
      class="MuiBox-root css-dtw560"
    >
      <img
        alt="Books cover"
        data-testid="BookGridCell.img"
        src="https://covers.openlibrary.org/b/id/13565208-M.jpg"
      />
    </div>
  </div>
</div>
`;
